nvflare/app_common/streamers/file_downloader.py

```diff
@@ -0,0 +1,230 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os.path
+import tempfile
+import uuid
+from typing import Any, List, Optional
+
+from nvflare.fuel.f3.cellnet.cell import Cell
+from nvflare.fuel.utils.validation_utils import check_positive_int
+
+from .obj_downloader import Consumer, ObjDownloader, Producer, ProduceRC, download_object
+
+DEFAULT_CHUNK_SIZE = 5 * 1024 * 1024
+
+"""
+This package implements file downloading capability based on the ObjDownloader framework.
+It provides implementation of the Producer and Consumer objects, required by ObjDownloader.
+"""
+
+
+class _StateKey:
+    RECEIVED_BYTES = "received_bytes"
+
+
+class _File:
+
+    def __init__(self, file_name):
+        """This is the "object" to be downloaded.
+
+        Args:
+            file_name: name of the file.
+        """
+        self.name = file_name
+        self.size = os.path.getsize(file_name)
+
+
+class _ChunkProducer(Producer):
+
+    def __init__(self, chunk_size=None):
+        Producer.__init__(self)
+        if not chunk_size:
+            chunk_size = DEFAULT_CHUNK_SIZE
+
+        check_positive_int("chunk_size", chunk_size)
+        self.chunk_size = chunk_size
+
+    def produce(self, ref_id: str, obj, state: dict, requester: str) -> (str, Any, dict):
+        assert isinstance(obj, _File)
+        received_bytes = 0
+        if state:
+            received_bytes = state.get(_StateKey.RECEIVED_BYTES, 0)
+
+        if not isinstance(received_bytes, int) or received_bytes < 0:
+            self.logger.error(f"bad {_StateKey.RECEIVED_BYTES} {received_bytes} from {requester}")
+            return ProduceRC.ERROR, None, None
+
+        if received_bytes >= obj.size:
+            # already done
+            return ProduceRC.EOF, None, None
+
+        num_bytes_to_send = min(self.chunk_size, obj.size - received_bytes)
+        with open(obj.name, "rb") as f:
+            f.seek(received_bytes)
```

Comment:
The seek() here is not necessary. It should always point to the current position.